man/jc.1
.TH jc 1 2026-10-05 1.26.1 "JSON Convert"
.SH NAME
\fBjc\fP \- JSON Convert JSONifies the output of many CLI tools, file-types,
and strings
.SH SYNOPSIS

Standard syntax:

.RS
COMMAND | \fBjc\fP [SLICE] [OPTIONS] PARSER

cat FILE | \fBjc\fP [SLICE] [OPTIONS] PARSER

echo STRING | \fBjc\fP [SLICE] [OPTIONS] PARSER
.RE

Magic syntax:

.RS
\fBjc\fP [SLICE] [OPTIONS] COMMAND

\fBjc\fP [SLICE] [OPTIONS] /proc/<path-to-procfile>
.RE

.SH DESCRIPTION
\fBjc\fP JSONifies the output of many CLI tools, file-types, and common strings
for easier parsing in scripts. \fBjc\fP accepts piped input from \fBSTDIN\fP and
outputs a JSON representation of the previous command's output to \fBSTDOUT\fP.
Alternatively, the "Magic" syntax can be used by prepending \fBjc\fP to the
command to be converted. Options can be passed to \fBjc\fP immediately before
the command is given. (Note: "Magic" syntax does not support shell builtins or
command aliases)

.SH OPTIONS
.B
Parsers:
.RS


.TP
.B
\fB--acpi\fP
`acpi` command parser

.TP
.B
\fB--airport\fP
`airport -I` command parser

.TP
.B
\fB--airport-s\fP
`airport -s` command parser

.TP
.B
\fB--amixer\fP
`amixer` command parser

.TP
.B
\fB--apt-cache-show\fP
`apt-cache show` command parser

.TP
.B
\fB--apt-get-sqq\fP
`apt-get -sqq` command parser

.TP
.B
\fB--arp\fP
`arp` command parser

.TP
.B
\fB--asciitable\fP
ASCII and Unicode table parser

.TP
.B
\fB--asciitable-m\fP
multi-line ASCII and Unicode table parser

.TP
.B
\fB--authorized-keys\fP
`~/.ssh/authorized_keys` file parser

.TP
.B
\fB--blkid\fP
`blkid` command parser

.TP
.B
\fB--bluetoothctl\fP
`bluetoothctl` command parser

.TP
.B
\fB--cbt\fP
`cbt` (Google Bigtable) command parser

.TP
.B
\fB--cef\fP
CEF string parser

.TP
.B
\fB--cef-s\fP
CEF string streaming parser

.TP
.B
\fB--certbot\fP
`certbot` command parser

.TP
.B
\fB--chage\fP
`chage --list` command parser

.TP
.B
\fB--cksum\fP
`cksum` and `sum` command parser

.TP
.B
\fB--clf\fP
Common and Combined Log Format file parser

.TP
.B
\fB--clf-s\fP
Common and Combined Log Format file streaming parser

.TP
.B
\fB--crontab\fP
`crontab` command and file parser

.TP
.B
\fB--crontab-u\fP
`crontab` file parser with user support

.TP
.B
\fB--csv\fP
CSV file parser

.TP
.B
\fB--csv-ih\fP
CSV implicit header file parser

.TP
.B
\fB--csv-s\fP
CSV file streaming parser

.TP
.B
\fB--csv-ih-s\fP
CSV implicit header file streaming parser

.TP
.B
\fB--curl-head\fP
`curl --head` command parser

.TP
.B
\fB--date\fP
`date` command parser

.TP
.B
\fB--datetime-iso\fP
ISO 8601 Datetime string parser

.TP
.B
\fB--debconf-show\fP
`debconf-show` command parser

.TP
.B
\fB--df\fP
`df` command parser

.TP
.B
\fB--dig\fP
`dig` command parser

.TP
.B
\fB--dir\fP
`dir` command parser

.TP
.B
\fB--dmidecode\fP
`dmidecode` command parser

.TP
.B
\fB--dpkg-l\fP
`dpkg -l` command parser

.TP
.B
\fB--du\fP
`du` command parser

.TP
.B
\fB--efibootmgr\fP
`efibootmgr` command parser

.TP
.B
\fB--email-address\fP
Email Address string parser

.TP
.B
\fB--env\fP
`env` command parser

.TP
.B
\fB--ethtool\fP
`ethtool` command parser

.TP
.B
\fB--file\fP
`file` command parser

.TP
.B
\fB--find\fP
`find` command parser

.TP
.B
\fB--findmnt\fP
`findmnt` command parser

.TP
.B
\fB--finger\fP
`finger` command parser

.TP
.B
\fB--free\fP
`free` command parser

.TP
.B
\fB--fstab\fP
`/etc/fstab` file parser

.TP
.B
\fB--getfacl\fP
`getfacl` command parser

.TP
.B
\fB--git-diff\fP
`git diff --name-status` command parser

.TP
.B
\fB--git-log\fP
`git log` command parser

.TP
.B
\fB--git-log-s\fP
`git log` command streaming parser

.TP
.B
\fB--git-ls-remote\fP
`git ls-remote` command parser

.TP
.B
\fB--gpg\fP
`gpg --with-colons` command parser

.TP
.B
\fB--group\fP
`/etc/group` file parser

.TP
.B
\fB--gshadow\fP
`/etc/gshadow` file parser

.TP
.B
\fB--hash\fP
`hash` command parser

.TP
.B
\fB--hashsum\fP
hashsum command parser (`md5sum`, `shasum`, etc.)

.TP
.B
\fB--hciconfig\fP
`hciconfig` command parser

.TP
.B
\fB--history\fP
`history` command parser

.TP
.B
\fB--host\fP
`host` command parser

.TP
.B
\fB--hosts\fP
`/etc/hosts` file parser

.TP
.B
\fB--http-headers\fP
HTTP headers parser

.TP
.B
\fB--id\fP
`id` command parser

.TP
.B
\fB--ifconfig\fP
`ifconfig` command parser

.TP
.B
\fB--iftop\fP
`iftop` command parser

.TP
.B
\fB--ini\fP
INI file parser

.TP
.B
\fB--ini-dup\fP
INI with duplicate key file parser

.TP
.B
\fB--iostat\fP
`iostat` command parser

.TP
.B
\fB--iostat-s\fP
`iostat` command streaming parser

.TP
.B
\fB--ip-address\fP
IPv4 and IPv6 Address string parser

.TP
.B
\fB--ipconfig\fP
`ipconfig` Windows command parser

.TP
.B
\fB--iptables\fP
`iptables` command parser

.TP
.B
\fB--ip-route\fP
`ip route` command parser

.TP
.B
\fB--iw-scan\fP
`iw dev [device] scan` command parser

.TP
.B
\fB--iwconfig\fP
`iwconfig` command parser

.TP
.B
\fB--jar-manifest\fP
Java MANIFEST.MF file parser

.TP
.B
\fB--jobs\fP
`jobs` command parser

.TP
.B
\fB--jwt\fP
JWT string parser

.TP
.B
\fB--kv\fP
Key/Value file and string parser

.TP
.B
\fB--kv-dup\fP
Key/Value with duplicate key file and string parser

.TP
.B
\fB--last\fP
`last` and `lastb` command parser

.TP
.B
\fB--ldd\fP
`ldd` command parser

.TP
.B
\fB--ls\fP
`ls` command parser

.TP
.B
\fB--ls-s\fP
`ls` command streaming parser

.TP
.B
\fB--lsattr\fP
`lsattr` command parser

.TP
.B
\fB--lsb-release\fP
`lsb_release` command parser

.TP
.B
\fB--lsblk\fP
`lsblk` command parser

.TP
.B
\fB--lsmod\fP
`lsmod` command parser

.TP
.B
\fB--lsof\fP
`lsof` command parser

.TP
.B
\fB--lspci\fP
`lspci -mmv` command parser

.TP
.B
\fB--lsusb\fP
`lsusb` command parser

.TP
.B
\fB--m3u\fP
M3U and M3U8 file parser

.TP
.B
\fB--mdadm\fP
`mdadm` command parser

.TP
.B
\fB--mount\fP
`mount` command parser

.TP
.B
\fB--mpstat\fP
`mpstat` command parser

.TP
.B
\fB--mpstat-s\fP
`mpstat` command streaming parser

.TP
.B
\fB--needrestart\fP
`needrestart -b` command parser

.TP
.B
\fB--netrc\fP
`.netrc` file parser

.TP
.B
\fB--netstat\fP
`netstat` command parser

.TP
.B
\fB--net-localgroup\fP
`net localgroup` command parser

.TP
.B
\fB--net-user\fP
`net user` command parser

.TP
.B
\fB--nmcli\fP
`nmcli` command parser

.TP
.B
\fB--nsd-control\fP
`nsd-control` command parser

.TP
.B
\fB--ntpq\fP
`ntpq -p` command parser

.TP
.B
\fB--openvpn\fP
openvpn-status.log file parser

.TP
.B
\fB--os-prober\fP
`os-prober` command parser

.TP
.B
\fB--os-release\fP
`/etc/os-release` file parser

.TP
.B
\fB--pacman\fP
`pacman` command parser

.TP
.B
\fB--passwd\fP
`/etc/passwd` file parser

.TP
.B
\fB--path\fP
POSIX path string parser

.TP
.B
\fB--path-list\fP
POSIX path list string parser

.TP
.B
\fB--pci-ids\fP
`pci.ids` file parser

.TP
.B
\fB--pgpass\fP
PostgreSQL password file parser

.TP
.B
\fB--pidstat\fP
`pidstat -H` command parser

.TP
.B
\fB--pidstat-s\fP
`pidstat -H` command streaming parser

.TP
.B
\fB--ping\fP
`ping` and `ping6` command parser

.TP
.B
\fB--ping-s\fP
`ping` and `ping6` command streaming parser

.TP
.B
\fB--pip-list\fP
`pip list` command parser

.TP
.B
\fB--pip-show\fP
`pip show` command parser

.TP
.B
\fB--pkg-index-apk\fP
Alpine Linux Package Index file parser

.TP
.B
\fB--pkg-index-deb\fP
Debian Package Index file parser

.TP
.B
\fB--plist\fP
PLIST file parser

.TP
.B
\fB--postconf\fP
`postconf -M` command parser

.TP
.B
\fB--proc\fP
`/proc/` file parser

.TP
.B
\fB--proc-buddyinfo\fP
`/proc/buddyinfo` file parser

.TP
.B
\fB--proc-cmdline\fP
`/proc/cmdline` file parser

.TP
.B
\fB--proc-consoles\fP
`/proc/consoles` file parser

.TP
.B
\fB--proc-cpuinfo\fP
`/proc/cpuinfo` file parser

.TP
.B
\fB--proc-crypto\fP
`/proc/crypto` file parser

.TP
.B
\fB--proc-devices\fP
`/proc/devices` file parser

.TP
.B
\fB--proc-diskstats\fP
`/proc/diskstats` file parser

.TP
.B
\fB--proc-filesystems\fP
`/proc/filesystems` file parser

.TP
.B
\fB--proc-interrupts\fP
`/proc/interrupts` file parser

.TP
.B
\fB--proc-iomem\fP
`/proc/iomem` file parser

.TP
.B
\fB--proc-ioports\fP
`/proc/ioports` file parser

.TP
.B
\fB--proc-loadavg\fP
`/proc/loadavg` file parser

.TP
.B
\fB--proc-locks\fP
`/proc/locks` file parser

.TP
.B
\fB--proc-meminfo\fP
`/proc/meminfo` file parser

.TP
.B
\fB--proc-modules\fP
`/proc/modules` file parser

.TP
.B
\fB--proc-mtrr\fP
`/proc/mtrr` file parser

.TP
.B
\fB--proc-pagetypeinfo\fP
`/proc/pagetypeinfo` file parser

.TP
.B
\fB--proc-partitions\fP
`/proc/partitions` file parser

.TP
.B
\fB--proc-slabinfo\fP
`/proc/slabinfo` file parser

.TP
.B
\fB--proc-softirqs\fP
`/proc/softirqs` file parser

.TP
.B
\fB--proc-stat\fP
`/proc/stat` file parser

.TP
.B
\fB--proc-swaps\fP
`/proc/swaps` file parser

.TP
.B
\fB--proc-uptime\fP
`/proc/uptime` file parser

.TP
.B
\fB--proc-version\fP
`/proc/version` file parser

.TP
.B
\fB--proc-vmallocinfo\fP
`/proc/vmallocinfo` file parser

.TP
.B
\fB--proc-vmstat\fP
`/proc/vmstat` file parser

.TP
.B
\fB--proc-zoneinfo\fP
`/proc/zoneinfo` file parser

.TP
.B
\fB--proc-driver-rtc\fP
`/proc/driver/rtc` file parser

.TP
.B
\fB--proc-net-arp\fP
`/proc/net/arp` file parser

.TP
.B
\fB--proc-net-dev\fP
`/proc/net/dev` file parser

.TP
.B
\fB--proc-net-dev-mcast\fP
`/proc/net/dev_mcast` file parser

.TP
.B
\fB--proc-net-if-inet6\fP
`/proc/net/if_inet6` file parser

.TP
.B
\fB--proc-net-igmp\fP
`/proc/net/igmp` file parser

.TP
.B
\fB--proc-net-igmp6\fP
`/proc/net/igmp6` file parser

.TP
.B
\fB--proc-net-ipv6-route\fP
`/proc/net/ipv6_route` file parser

.TP
.B
\fB--proc-net-netlink\fP
`/proc/net/netlink` file parser

.TP
.B
\fB--proc-net-netstat\fP
`/proc/net/netstat` file parser

.TP
.B
\fB--proc-net-packet\fP
`/proc/net/packet` file parser

.TP
.B
\fB--proc-net-protocols\fP
`/proc/net/protocols` file parser

.TP
.B
\fB--proc-net-route\fP
`/proc/net/route` file parser

.TP
.B
\fB--proc-net-tcp\fP
`/proc/net/tcp` and `/proc/net/tcp6` file parser

.TP
.B
\fB--proc-net-unix\fP
`/proc/net/unix` file parser

.TP
.B
\fB--proc-pid-fdinfo\fP
`/proc/<pid>/fdinfo/<fd>` file parser

.TP
.B
\fB--proc-pid-io\fP
`/proc/<pid>/io` file parser

.TP
.B
\fB--proc-pid-maps\fP
`/proc/<pid>/maps` file parser

.TP
.B
\fB--proc-pid-mountinfo\fP
`/proc/<pid>/mountinfo` file parser

.TP
.B
\fB--proc-pid-numa-maps\fP
`/proc/<pid>/numa_maps` file parser

.TP
.B
\fB--proc-pid-smaps\fP
`/proc/<pid>/smaps` file parser

.TP
.B
\fB--proc-pid-stat\fP
`/proc/<pid>/stat` file parser

.TP
.B
\fB--proc-pid-statm\fP
`/proc/<pid>/statm` file parser

.TP
.B
\fB--proc-pid-status\fP
`/proc/<pid>/status` file parser

.TP
.B
\fB--ps\fP
`ps` command parser

.TP
.B
\fB--resolve-conf\fP
`/etc/resolve.conf` file parser

.TP
.B
\fB--route\fP
`route` command parser

.TP
.B
\fB--route-print\fP
`route print` command parser

.TP
.B
\fB--rpm-qi\fP
`rpm -qi` command parser

.TP
.B
\fB--rsync\fP
`rsync` command parser

.TP
.B
\fB--rsync-s\fP
`rsync` command streaming parser

.TP
.B
\fB--semver\fP
Semantic Version string parser

.TP
.B
\fB--sfdisk\fP
`sfdisk` command parser

.TP
.B
\fB--shadow\fP
`/etc/shadow` file parser

.TP
.B
\fB--srt\fP
SRT file parser

.TP
.B
\fB--ss\fP
`ss` command parser

.TP
.B
\fB--ssh-conf\fP
`ssh` config file and `ssh -G` command parser

.TP
.B
\fB--sshd-conf\fP
`sshd` config file and `sshd -T` command parser

.TP
.B
\fB--stat\fP
`stat` command parser

.TP
.B
\fB--stat-s\fP
`stat` command streaming parser

.TP
.B
\fB--swapon\fP
`swapon` command parser

.TP
.B
\fB--sysctl\fP
`sysctl` command parser

.TP
.B
\fB--syslog\fP
Syslog RFC 5424 string parser

.TP
.B
\fB--syslog-s\fP
Syslog RFC 5424 string streaming parser

.TP
.B
\fB--syslog-bsd\fP
Syslog RFC 3164 string parser

.TP
.B
\fB--syslog-bsd-s\fP
Syslog RFC 3164 string streaming parser

.TP
.B
\fB--systemctl\fP
`systemctl` command parser

.TP
.B
\fB--systemctl-lj\fP
`systemctl list-jobs` command parser

.TP
.B
\fB--systemctl-ls\fP
`systemctl list-sockets` command parser

.TP
.B
\fB--systemctl-luf\fP
`systemctl list-unit-files` command parser

.TP
.B
\fB--systeminfo\fP
`systeminfo` command parser

.TP
.B
\fB--time\fP
`/usr/bin/time` command parser

.TP
.B
\fB--timedatectl\fP
`timedatectl status` command parser

.TP
.B
\fB--timestamp\fP
Unix Epoch Timestamp string parser

.TP
.B
\fB--toml\fP
TOML file parser

.TP
.B
\fB--top\fP
`top -b` command parser

.TP
.B
\fB--top-s\fP
`top -b` command streaming parser

.TP
.B
\fB--tracepath\fP
`tracepath` and `tracepath6` command parser

.TP
.B
\fB--traceroute\fP
`traceroute` and `traceroute6` command parser

.TP
.B
\fB--traceroute-s\fP
`traceroute` and `traceroute6` command streaming parser

.TP
.B
\fB--tsv\fP
TSV file parser

.TP
.B
\fB--tsv-ih\fP
TSV implicit header file parser

.TP
.B
\fB--tsv-s\fP
TSV file streaming parser

.TP
.B
\fB--tsv-ih-s\fP
TSV implicit header file streaming parser

.TP
.B
\fB--tune2fs\fP
`tune2fs -l` command parser

.TP
.B
\fB--typeset\fP
`typeset` and `declare` command parser

.TP
.B
\fB--udevadm\fP
`udevadm info` command parser

.TP
.B
\fB--ufw\fP
`ufw status` command parser

.TP
.B
\fB--ufw-appinfo\fP
`ufw app info [application]` command parser

.TP
.B
\fB--uname\fP
`uname -a` command parser

.TP
.B
\fB--update-alt-gs\fP
`update-alternatives --get-selections` command parser

.TP
.B
\fB--update-alt-q\fP
`update-alternatives --query` command parser

.TP
.B
\fB--upower\fP
`upower` command parser

.TP
.B
\fB--upsc\fP
`upsc` command parser

.TP
.B
\fB--uptime\fP
`uptime` command parser

.TP
.B
\fB--url\fP
URL string parser

.TP
.B
\fB--ver\fP
Version string parser

.TP
.B
\fB--veracrypt\fP
`veracrypt` command parser

.TP
.B
\fB--vmstat\fP
`vmstat` command parser

.TP
.B
\fB--vmstat-s\fP
`vmstat` command streaming parser

.TP
.B
\fB--w\fP
`w` command parser

.TP
.B
\fB--wc\fP
`wc` command parser

.TP
.B
\fB--wg-show\fP
`wg show` command parser

.TP
.B
\fB--who\fP
`who` command parser

.TP
.B
\fB--x509-cert\fP
X.509 PEM and DER certificate file parser

.TP
.B
\fB--x509-crl\fP
X.509 PEM and DER certificate revocation list file parser

.TP
.B
\fB--x509-csr\fP
X.509 PEM and DER certificate request file parser

.TP
.B
\fB--xml\fP
XML file parser

.TP
.B
\fB--xrandr\fP
`xrandr` command parser

.TP
.B
\fB--yaml\fP
YAML file parser

.TP
.B
\fB--zipinfo\fP
`zipinfo` command parser

.TP
.B
\fB--zpool-iostat\fP
`zpool iostat` command parser

.TP
.B
\fB--zpool-status\fP
`zpool status` command parser


.RE
.PP
.B
Options:
.RS

.TP
.B
\fB-a\fP, \fB--about\fP
About \fBjc\fP (JSON or YAML output)
.TP
.B
\fB-C\fP, \fB--force-color\fP
Force color output even when using pipes (overrides \fB-m\fP and the
\fBNO_COLOR\fP env variable)
.TP
.B
\fB-d\fP, \fB--debug\fP
Debug - show traceback (use \fB-dd\fP for verbose traceback)
.TP
.B
\fB-h\fP, \fB--help\fP
Help (\fB--help --parser_name\fP for parser documentation). Use twice to show
hidden parsers (e.g. \fB-hh\fP). Use thrice to show parser categories (e.g. \fB-hhh\fP).
.TP
.B
\fB-m\fP, \fB--monochrome\fP
Monochrome output
.TP
.B
\fB-M\fP, \fB--meta-out\fP
Add metadata to output including timestamp, parser name, magic command, magic
command exit code, etc.
.TP
.B
\fB-p\fP, \fB--pretty\fP
Pretty print output
.TP
.B
\fB-q\fP, \fB--quiet\fP
Quiet mode. Suppresses parser warning messages (use \fB-qq\fP to ignore streaming
parser errors)
.TP
.B
\fB-r\fP, \fB--raw\fP
Raw output. Provides more literal output, typically with string values and no
additional semantic processing
.TP
.B
\fB-s\fP, \fB--slurp\fP
Slurp multiple lines into an array. (use \fB-hhh\fP to find compatible parsers)
.TP
.B
\fB-u\fP, \fB--unbuffer\fP
Unbuffer output (useful for slow streaming data with streaming parsers)
.TP
.B
\fB-v\fP, \fB--version\fP
Version information
.TP
.B
\fB-y\fP, \fB--yaml-out\fP
YAML output
.TP
.B
\fB-B\fP, \fB--bash-comp\fP
Generate Bash shell completion script
.TP
.B
\fB-Z\fP, \fB--zsh-comp\fP
Generate Zsh shell completion script
.TP
.B
\fB-X\fP, \fB--xonsh-comp\fP
Generate Xonsh shell completion script

.RE
.PP
.B
Slice:
.RS
Line slicing is supported using the \fBSTART:STOP\fP syntax similar to Python
slicing. This allows you to skip lines at the beginning and/or end of the
\fBSTDIN\fP input you would like \fBjc\fP to convert.

\fBSTART\fP and \fBSTOP\fP can be positive or negative integers or blank and
allow you to specify how many lines to skip and how many lines to process.
Positive and blank slices are the most memory efficient. Any negative
integers in the slice will use more memory.

For example, to skip the first and last line of the following text, you
could express the slice in a couple ways:

.RS
.nf
$ cat table.txt
      ### We want to skip this header ###
          col1       col2
          foo        1
          bar        2
      ### We want to skip this footer ###
$ cat table.txt | jc 1:-1 --asciitable
[{"col1":"foo","col2":"1"},{"col1":"bar","col2":"2"}]
$ cat table.txt | jc 1:4 --asciitable
[{"col1":"foo","col2":"1"},{"col1":"bar","col2":"2"}]
.fi
.RE

In this example \fB1:-1\fP and \fB1:4\fP line slices provide the same output.

When using positive integers the index location of \fBSTOP\fP is non-inclusive.
Positive slices count from the first line of the input toward the end
starting at \fB0\fP as the first line. Negative slices count from the last line
toward the beginning starting at \fB-1\fP as the last line. This is also the way
Python's slicing feature works.

Here is a breakdown of line slice options:

.TP
.B
\fBSTART:STOP\fP
lines \fBSTART\fP through \fBSTOP - 1\fP
.TP
.B
\fBSTART:\fP
lines \fBSTART\fP through the rest of the output
.TP
.B
\fB:STOP\fP
lines from the beginning through \fBSTOP - 1\fP
.TP
.B
\fB-START:STOP\fP
\fBSTART\fP lines from the end through \fBSTOP - 1\fP
.TP
.B
\fBSTART:-STOP\fP
lines \fBSTART\fP through \fBSTOP\fP lines from the end
.TP
.B
\fB-START:-STOP\fP
\fBSTART\fP lines from the end through \fBSTOP\fP lines from the end
.TP
.B
\fB-START:\fP
\fBSTART\fP lines from the end through the rest of the output
.TP
.B
\fB:-STOP\fP
lines from the beginning through \fBSTOP\fP lines from the end
.TP
.B
\fB:\fP
all lines

.SH SLURP
Some parsers support multi-item input and can output an array of results in a
single pass. Slurping works for string parsers that accept a single line of
input. (e.g. \fBurl\fP and \fBip-address\fP) To see a list of parsers that support
the \fB--slurp\fP option, use \fBjc -hhh\fP.

For example, you can send a file with multiple IP addresses (one per line) to
\fBjc\fP with the \fB--slurp\fP option and an array of results will output:

.RS
.nf
$ cat ip-addresses.txt | jc --slurp --ip-address
[<multiple output objects>]
.fi
.RE

The magic syntax for \fB/proc\fP files automatically supports slurping of multiple
files (no need to use the \fB--slurp\fP option). For example, you can convert many
PID files at once:

.RS
.nf
$ jc /proc/*/status
[<multiple output objects>]
.fi
.RE

When the \fB/proc\fP magic syntax is used and multiple files are selected, an
additional \fB_file\fP field is inserted in the output so it is easier to tell what
file each output object refers to.

Finally, the \fB--meta-out\fP option can be used in conjunction with slurped output.
In this case, the slurped output is wrapped in an object with the following
structure:

.RS
.nf
{
  "result": [<multiple output objects>],
  "_jc_meta": {
    "parser": "url",
    "timestamp": 1706235558.654576,
    "slice_start": null,
    "slice_end": null,
    "input_list": [
      "http://www.google.com",
      "https://www.apple.com",
      "https://www.microsoft.com"
    ]
  }
}
.fi
.RE

With \fB--meta-out\fP, \fBinput_list\fP contains a list of inputs (actual input strings
or \fB/proc\fP filenames) so you can identify which output object relates to each
input string or \fB/proc\fP filename.

.SH EXIT CODES
Any fatal errors within \fBjc\fP will generate an exit code of \fB100\fP,
otherwise the exit code will be \fB0\fP.

When using the "magic" syntax (e.g. \fBjc ifconfig eth0\fP), \fBjc\fP will store
the exit code of the program being parsed and add it to the \fBjc\fP exit code.
This way it is easier to determine if an error was from the parsed program or
\fBjc\fP.

Consider the following examples using \fBifconfig\fP:

.RS
ifconfig exit code = \fB0\fP, jc exit code = \fB0\fP, combined exit code = \fB0\fP (no errors)

ifconfig exit code = \fB1\fP, jc exit code = \fB0\fP, combined exit code = \fB1\fP (error in ifconfig)


CHANGELOG
jc changelog

20261005 v1.26.1
- Fix `fstab` parser to ignore whitespace-only lines
- Fix `group` parser to ignore whitespace-only lines
- Fix `gshadow` parser to ignore whitespace-only lines
- Fix `passwd` parser to ignore whitespace-only lines
- Fix `plist` parser for deprecation warning for codecs.open()
- Fix `shadow` parser to ignore whitespace-only lines

20260919 v1.26.0
- Add `authorized-keys` file parser for `~/.ssh/authorized_keys` files
- Add `getfacl` command parser
- Add `git diff --name-status` command parser
- Add `iftop` command parser (note this parser was added long ago but never
  registered so it was effectively hidden)
- Add `ldd` command parser
- Add `.netrc` file parser
- Add `upsc` command parser
- Add xonsh shell completion
- Enhance `git-log` standard and streaming parsers with added `insertions`
  and `deletions` fields under `file_stats` when using `--numstat`. Also added
  `--patch` output support
- Enhance `nmcli` parser to support `nmcli -t device wifi list`
- Enhance `typeset` parser to support ksh and zsh output
- Enhance `typeset` parser to handle ANSI-C quoting: `$'...'`
- Enhance `ls` and `ls-s` standard and streaming parsers to support device
  files and added support for long-iso dates
- Enhance `lsusb` parser to support USB Audio and MIDI class interface and
  endpoint descriptors (AudioControl, AudioStreaming, MIDIStreaming)
- Simplify `lsusb` parser internals to build nested output in a single pass
  Note: this is a complete rewrite (v2.0) and it is possible some output may
  deviate from the previous version. It is believed this version is more
  accurate and will be easier to maintain going forward.
- Fix `curl-head` parser to ignore whitespace-only lines
- Fix `email-address` parser to gracefully handle invalid email addresses
- Fix `fstab` parser for lines that omit the fs_freq and fs_passno fields
- Fix `hosts` parser to ignore whitespace only lines
- Fix `http-headers` parser to ignore whitespace-only lines
- Fix `ip-route` parser to handle whitespace-only input gracefully
- Fix `pgpass` parser handling of escaped characters
- Fix `proc-modules` parser to support additional taint status column
- Fix `ss` parser for process names with spaces and other special characters
  and correctly parse the options field when it is padded
- Fix `traceroute` parser to handle missing hop information gracefully
- Fix `typeset` parser to allow scalar values starting with the `[` character
- Fix `ufw` parser to handle `(log-all)` messages
- Fix path traversal vulnerability with magic syntax
- Fix tests for Debian build
- Fix for broken locale behavior (permanently changing to user's locale) when
  unparsable datetimes are handled
- Fix timestamp function for 12 hour format found in some `date` command output

20260617 v1.25.7
- Add `csv-ih`, `tsv`, and `tsv-ih` file parsers for CSV files without headers,
  TSV files, and TSV files without headers. Note that these are standard and
  not streaming parsers so the entire CSV/TSV will be loaded into memory. For
  lazy loading equivalents, see below.
- Add `csv-ih-s`, `tsv-s`, and `tsv-ih-s` streaming file parsers for CSV files
  without headers, TSV files, and TSV files without headers. These streaming
  parsers will lazily load the data into memory and may be more suitable for
  extremely large files.
- Add `typeset` and `declare` Bash internal command parser to convert variables
  simple arrays, and associative arrays along with object metadata
- Enhance `pip-show` command parser to add `-f` show files support
- Enhance `rsync` and `rsync-s` parsers to add `--stats` or `--info=stats[1-3]` fields
- Fix `certbot` command parser to allow parsing of named certs
- Fix `dir` command parser for incorrect stripping of the `D:` drive letter
- Fix `hashsum` command parser to correctly parse the `mode` indicator
- Fix `ifconfig` command parser for incorrect stripping of leading zeros in some hex numbers
- Fix `iptables` command parser when Target is blank and verbose output is used
- Fix `lsattr` command parser for filenames with spaces (newlines in filenames are
  still not supported)
- Fix `proc-pid-smaps` proc parser when unknown VmFlags are output
- Fix `jc.utils.timestamp` function and tests for behavior change in python 3.15
- Fix `ufw` command parser to support rule logging. A new boolean `log` field is introduced

20251012 v1.25.6
- Add `net-localgroup` Windows command parser
- Add `net-user` Windows command parser
- Add `route-print` Windows command parser
- Add `traceroute-s` streaming command parser
- Add `max_hops` and `data_bytes` fields to `traceroute` command parser
- Add `x509-crl` file parser to support Certificate Revocation List PEM and DER files
- Add `yay` as a magic command for the `pacman` command parser
- Fix `bluetoothctl` command parser to support output with the `cable_pairing` attribute
- Fix `nmcli` command parser to support blank `team.config` JSON value and `team-port.config` JSON value
- Fix `top` command parsers to correct memory size field parsing. Several new unit
  and byte conversion fields have been added
- Fix `who` command parser to support the `process` field on Debian13

20250503 v1.25.5
- Add `amixer` command parser
- Enhance `iptables` command parser to add default policy statistics fields
- Fix `bluetoothctl` parser failing to parse controllers with power state prop
- Fix `lsblk` command parser to support multiple mountpoints. Also, added
  byte conversions for size fields.
- Fix `nmcli` command parser to support `team.config` JSON field
- Fix `time` command parser for output that does not contain centiseconds
- Fix `x509-cert` parser to handle IDNA2008 encoded email addresses with a warning
- Fix typing for upcoming python v3.14
- Fix timezone setting for tests to support minimal chrooted builds

20241125 v1.25.4
- Add `ipconfig` command parser (`ipconfig` for Windows)
- Add `pacman` command parser
- Add `wg show` command parser

jc/jc.py
#!/usr/bin/env python3
"""jc - JSON CLI output utility

Main input module
"""

import sys
import signal
import json
from jc.utils import *
import jc.parsers.arp
import jc.parsers.df
import jc.parsers.dig
import jc.parsers.env
import jc.parsers.free
import jc.parsers.history
import jc.parsers.ifconfig
import jc.parsers.iptables
import jc.parsers.jobs
import jc.parsers.ls
import jc.parsers.lsblk
import jc.parsers.lsmod
import jc.parsers.lsof
import jc.parsers.mount
import jc.parsers.netstat
import jc.parsers.ps
import jc.parsers.route
import jc.parsers.uname
import jc.parsers.uptime
import jc.parsers.w


def main():
    signal.signal(signal.SIGINT, ctrlc)

    if sys.stdin.isatty():
        helptext('missing piped data')
        exit()

    data = sys.stdin.read()
    pretty = False
    quiet = False
    raw = False

    # options
    if '-p' in sys.argv:
        pretty = True

    if '-q' in sys.argv:
        quiet = True

    if '-r' in sys.argv:
        raw = True

    # parsers
    if '--arp' in sys.argv:
        result = jc.parsers.arp.parse(data, raw=raw, quiet=quiet)

    elif '--df' in sys.argv:
        result = jc.parsers.df.parse(data, raw=raw, quiet=quiet)

    elif '--dig' in sys.argv:
        result = jc.parsers.dig.parse(data, raw=raw, quiet=quiet)

    elif '--env' in sys.argv:
        result = jc.parsers.env.parse(data, raw=raw, quiet=quiet)

    elif '--free' in sys.argv:
        result = jc.parsers.free.parse(data, raw=raw, quiet=quiet)

    elif '--history' in sys.argv:
        result = jc.parsers.history.parse(data, raw=raw, quiet=quiet)

    elif '--ifconfig' in sys.argv:
        result = jc.parsers.ifconfig.parse(data, raw=raw, quiet=quiet)

    elif '--iptables' in sys.argv:
        result = jc.parsers.iptables.parse(data, raw=raw, quiet=quiet)

    elif '--jobs' in sys.argv:
        result = jc.parsers.jobs.parse(data, raw=raw, quiet=quiet)

    elif '--ls' in sys.argv:
        result = jc.parsers.ls.parse(data, raw=raw, quiet=quiet)

    elif '--lsblk' in sys.argv:
        result = jc.parsers.lsblk.parse(data, raw=raw, quiet=quiet)

    elif '--lsmod' in sys.argv:
        result = jc.parsers.lsmod.parse(data, raw=raw, quiet=quiet)

    elif '--lsof' in sys.argv:
        result = jc.parsers.lsof.parse(data, raw=raw, quiet=quiet)

    elif '--mount' in sys.argv:
        result = jc.parsers.mount.parse(data, raw=raw, quiet=quiet)

    elif '--netstat' in sys.argv:
        result = jc.parsers.netstat.parse(data, raw=raw, quiet=quiet)

    elif '--ps' in sys.argv:
        result = jc.parsers.ps.parse(data, raw=raw, quiet=quiet)

    elif '--route' in sys.argv:
        result = jc.parsers.route.parse(data, raw=raw, quiet=quiet)

    elif '--uname' in sys.argv:
        result = jc.parsers.uname.parse(data, raw=raw, quiet=quiet)

    elif '--uptime' in sys.argv:
        result = jc.parsers.uptime.parse(data, raw=raw, quiet=quiet)

    elif '--w' in sys.argv:
        result = jc.parsers.w.parse(data, raw=raw, quiet=quiet)

    else:
        helptext('missing or incorrect arguments')
        exit()

    # output resulting dictionary as json
    if pretty:
        print(json.dumps(result, indent=2))
    else:
        print(json.dumps(result))


if __name__ == '__main__':
    main()
